pep-0544.txt

```diff
@@ -0,0 +1,968 @@
+  (and subclass instances), and similarly for other built-in classes. Making
+  ``int`` a structural type wouldn't be safe without major changes to the
+  Python runtime, which won't happen.
+
+
+Support optional protocol members
+---------------------------------
+
+We can come up with examples where it would be handy to be able to say
+that a method or data attribute does not need to be present in a class
+implementing a protocol, but if it is present, it must conform to a specific
+signature or type. One could use a ``hasattr()`` check to determine whether
+they can use the attribute on a particular instance.
+
+Languages such as TypeScript have similar features and
+apparently they are pretty commonly used. The current realistic potential
+use cases for protocols in Python don't require these. In the interest
+of simplicity, we propose to not support optional methods or attributes.
+We can always revisit this later if there is an actual need.
+
+
+Make protocols interoperable with other approaches
+--------------------------------------------------
+
+The protocols as described here are basically a minimal extension to
+the existing concept of ABCs. We argue that this is the way they should
+be understood, instead of as something that replaces Zope interfaces,
+for example. Attempting such interoperabilities will significantly
+complicate both the concept and the implementation.
+
+
+Use assignments to check explicitly that a class implements a protocol
+----------------------------------------------------------------------
+
+In Go language the explicit checks for implementation are performed
+via dummy assignments [golang]_. Such a way is also possible with the
+current proposal. Example::
+
+  class A:
+      def __len__(self) -> float:
+          return ...
+
+  _: Sized = A()  # Error: A.__len__ doesn't conform to 'Sized'
+                  # (Incompatible return type 'float')
+
+This approach moves the check away from
+the class definition and it almost requires a comment as otherwise
+the code probably would not make any sense to an average reader
+-- it looks like dead code. Besides, in the simplest form it requires one
+to construct an instance of ``A``, which could be problematic if this requires
+accessing or allocating some resources such as files or sockets.
+We could work around the latter by using a cast, for example, but then
+the code would be ugly. Therefore we discourage the use of this pattern.
+
+
+Support ``isinstance()`` checks by default
+------------------------------------------
+
+The problem with this is instance checks could be unreliable, except for
+situations where there is a common signature convention such as ``Iterable``.
+Another potentially problematic case is assignment of attributes
+*after* instantiation::
+
+  class P(Protocol):
+      x: int
+
+  class C:
+      def initialize(self) -> None:
+          self.x = 0
+
+  c = C()
+  isinstance(c1, P)  # False
+  c.initialize()
+  isinstance(c, P)  # True
```

Comment:
This is not surprising to me and at runtime I can see this as being a useful feature. You can imagine:
```
class Client(Protocol):
    ...

class Connected(Protocol):
    connection: socket


...  # later in the code
isinstance(someClient, Connected)
```

Maybe we don't want to encourage this programming style but it definitely descibes some real world code, sigh.